From cd6a0bec8d161822c7c9e67b39b126ed91b35c98 Mon Sep 17 00:00:00 2001
From: Zhongteng Gui <dragon-archer@outlook.com>
Date: Sun, 18 Oct 2026 22:42:45 +0800
Subject: [PATCH 45/N] Cygwin: fix build under ISO C++ 20

winsup/cygwin/local_includes/pinfo.h: error: ISO C++ says that these are
ambiguous, even though the worst conversion for the first is better than
the worst conversion for the second: [-Werror]

winsup/cygwin/spawn.cc, winsup/cygwin/mm/malloc.cc,
winsup/cygserver/sysv_shm.cc: -Werror=unused-but-set-variable=

winsup/cygwin/strfuncs.cc: error: cannot bind non-const lvalue reference
of type 'char8_t&' to a value of type 'unsigned char'

Signed-off-by: Zhongteng Gui <dragon-archer@outlook.com>
---
 winsup/cygserver/sysv_shm.cc         | 16 ++++++++--------
 winsup/cygwin/local_includes/pinfo.h |  7 ++-----
 winsup/cygwin/mm/malloc.cc           |  4 ++++
 winsup/cygwin/spawn.cc               |  2 --
 winsup/cygwin/strfuncs.cc            |  6 +++---
 5 files changed, 17 insertions(+), 18 deletions(-)

diff --git a/winsup/cygserver/sysv_shm.cc b/winsup/cygserver/sysv_shm.cc
index 4578c53..4c3ea34 100644
--- a/winsup/cygserver/sysv_shm.cc
+++ b/winsup/cygserver/sysv_shm.cc
@@ -342,7 +342,6 @@ kern_shmat(struct thread *td, int shmid, const void *shmaddr, int shmflg)
 	struct shm_handle *shm_handle;
 #endif
 	vm_offset_t attach_va;
-	vm_prot_t prot;
 	vm_size_t size;
 #ifndef __CYGWIN__
 	int rv;
@@ -379,13 +378,6 @@ kern_shmat(struct thread *td, int shmid, const void *shmaddr, int shmflg)
 		goto done2;
 	}
 	size = round_page(shmseg->shm_segsz);
-#ifdef VM_PROT_READ_IS_EXEC
-	prot = VM_PROT_READ | VM_PROT_EXECUTE;
-#else
-	prot = VM_PROT_READ;
-#endif
-	if ((shmflg & SHM_RDONLY) == 0)
-		prot |= VM_PROT_WRITE;
 	flags = MAP_ANON | MAP_SHARED;
 	debug_printf ("shmaddr: %x, shmflg: %x", shmaddr, shmflg);
 #ifdef __CYGWIN__
@@ -395,6 +387,14 @@ kern_shmat(struct thread *td, int shmid, const void *shmaddr, int shmflg)
 	   returns memory which isn't aligned to SHMLBA.  Go figure!  */
 	attach_va = (vm_offset_t)shmaddr;
 #else
+	vm_prot_t prot;
+#ifdef VM_PROT_READ_IS_EXEC
+	prot = VM_PROT_READ | VM_PROT_EXECUTE;
+#else
+	prot = VM_PROT_READ;
+#endif
+	if ((shmflg & SHM_RDONLY) == 0)
+		prot |= VM_PROT_WRITE;
 	if (shmaddr) {
 		flags |= MAP_FIXED;
 		if (shmflg & SHM_RND) {
diff --git a/winsup/cygwin/local_includes/pinfo.h b/winsup/cygwin/local_includes/pinfo.h
index 6f817de..947873a 100644
--- a/winsup/cygwin/local_includes/pinfo.h
+++ b/winsup/cygwin/local_includes/pinfo.h
@@ -203,7 +203,7 @@ public:
   class cygthread *wait_thread;
 
   void init (pid_t, DWORD, HANDLE);
-  pinfo (_pinfo *x = NULL)
+  explicit pinfo (_pinfo *x = NULL)
   : pinfo_minimal (), destroy (false), winpid_hdl (NULL), procinfo (x),
     waiter_ready (false), wait_thread (NULL) {}
   pinfo (pid_t n, DWORD flag = 0)
@@ -227,10 +227,7 @@ public:
   void maybe_set_exit_code_from_windows ();
   void set_exit_code (DWORD n);
   _pinfo *operator -> () const {return procinfo;}
-  int operator == (pinfo *x) const {return x->procinfo == procinfo;}
-  int operator == (pinfo &x) const {return x.procinfo == procinfo;}
-  int operator == (_pinfo *x) const {return x == procinfo;}
-  int operator == (void *x) const {return procinfo == x;}
+  bool operator == (const pinfo &x) const {return x.procinfo == procinfo;}
   _pinfo *operator * () const {return procinfo;}
   operator _pinfo * () const {return procinfo;}
   int operator !() const {return !procinfo;}
diff --git a/winsup/cygwin/mm/malloc.cc b/winsup/cygwin/mm/malloc.cc
index 8a1fc25..d32b3ba 100644
--- a/winsup/cygwin/mm/malloc.cc
+++ b/winsup/cygwin/mm/malloc.cc
@@ -3992,7 +3992,9 @@ static void add_segment(mstate m, char* tbase, size_t tsize, flag_t mmapped) {
   msegmentptr ss = (msegmentptr)(chunk2mem(sp));
   mchunkptr tnext = chunk_plus_offset(sp, ssize);
   mchunkptr p = tnext;
+#if DEBUG
   int nfences = 0;
+#endif
 
   /* reset top to new space */
   init_top(m, (mchunkptr)tbase, tsize - TOP_FOOT_SIZE);
@@ -4010,7 +4012,9 @@ static void add_segment(mstate m, char* tbase, size_t tsize, flag_t mmapped) {
   for (;;) {
     mchunkptr nextp = chunk_plus_offset(p, SIZE_T_SIZE);
     p->head = FENCEPOST_HEAD;
+#if DEBUG
     ++nfences;
+#endif
     if ((char*)(&(nextp->head)) < old_end)
       p = nextp;
     else
diff --git a/winsup/cygwin/spawn.cc b/winsup/cygwin/spawn.cc
index 15a121a..b0eeacf 100644
--- a/winsup/cygwin/spawn.cc
+++ b/winsup/cygwin/spawn.cc
@@ -338,7 +338,6 @@ child_info_spawn::worker (const char *prog_arg, const char *const *argv,
   int c_flags;
 
   STARTUPINFOW si = {};
-  int looped = 0;
 
   fhandler_termios::spawn_worker term_spawn_worker;
 
@@ -895,7 +894,6 @@ child_info_spawn::worker (const char *prog_arg, const char *const *argv,
 	    {
 	      if (!proc_retry (pi.hProcess))
 		{
-		  looped++;
 		  goto loop;
 		}
 	      close_all_files (true);
diff --git a/winsup/cygwin/strfuncs.cc b/winsup/cygwin/strfuncs.cc
index 0cf41ce..85427a0 100644
--- a/winsup/cygwin/strfuncs.cc
+++ b/winsup/cygwin/strfuncs.cc
@@ -211,9 +211,9 @@ c8rtomb (char *s, char8_t c8, mbstate_t *ps)
     {
       /* We already collected something... */
       int idx = ps->__count & 0x3;
-      char8_t &c1 = ps->__value.__wchb[0];
-      char8_t &c2 = ps->__value.__wchb[1];
-      char8_t &c3 = ps->__value.__wchb[2];
+      unsigned char &c1 = ps->__value.__wchb[0];
+      unsigned char &c2 = ps->__value.__wchb[1];
+      unsigned char &c3 = ps->__value.__wchb[2];
 
       switch (idx)
 	{
